db/migrate/20190617191109_move_awx_credentials_to_authentications.rb

```diff
@@ -0,0 +1,143 @@
+        awx_info = awx_credential_info(auth.manager_ref)
+        update_authentication(auth, awx_info)
+      end
+    end
+  rescue PG::ConnectionBad
+    say("awx database is unreachable, credentials cannot be migrated")
+  end
+
+  private
+
+  def embedded_ansible_authentications
+    types = %w[
+      ManageIQ::Providers::EmbeddedAnsible::AutomationManager::Credential
+      ManageIQ::Providers::EmbeddedAnsible::AutomationManager::CloudCredential
+      ManageIQ::Providers::EmbeddedAnsible::AutomationManager::MachineCredential
+      ManageIQ::Providers::EmbeddedAnsible::AutomationManager::NetworkCredential
+      ManageIQ::Providers::EmbeddedAnsible::AutomationManager::ScmCredential
+      ManageIQ::Providers::EmbeddedAnsible::AutomationManager::VaultCredential
+      ManageIQ::Providers::EmbeddedAnsible::AutomationManager::AmazonCredential
+      ManageIQ::Providers::EmbeddedAnsible::AutomationManager::AzureCredential
+      ManageIQ::Providers::EmbeddedAnsible::AutomationManager::GoogleCredential
+      ManageIQ::Providers::EmbeddedAnsible::AutomationManager::OpenstackCredential
+      ManageIQ::Providers::EmbeddedAnsible::AutomationManager::RhvCredential
+      ManageIQ::Providers::EmbeddedAnsible::AutomationManager::VmwareCredential
+    ]
+    Authentication.in_my_region.where(:type => types)
+  end
+
+  def awx_credential_info(awx_id)
+    cred_info = awx_connection.async_exec("SELECT inputs FROM main_credential WHERE id = $1::BIGINT", [awx_id]).first
+    # in case there is no matching credential on the awx side
+    return {} unless cred_info
+
+    JSON.parse(cred_info["inputs"])
+  end
+
+  def awx_connection
+    @awx_connection ||= PG::Connection.new(ApplicationRecord.connection.raw_connection.conninfo_hash.merge(:dbname => "awx").delete_blanks)
+  end
+
+  def update_authentication(auth, awx_info)
+    auth.options = auth.options.slice(*OPTIONS_FIELDS.map(&:to_sym)).presence
+
+    awx_info.each do |k, v|
+      if OPTIONS_FIELDS.include?(k)
+        auth.options ||= {}
+        auth.options[k.to_sym] = v
+        next
+      end
+
+      authentication_attribute = FIELD_MAP[k]
+      decrypted_value          = decrypted_awx_value(k, v, auth.manager_ref)
+      new_value                = ENCRYPTED_ATTRIBUTES.include?(authentication_attribute) ? ManageIQ::Password.encrypt(decrypted_value) : decrypted_value
+
+      if authentication_attribute
+        auth.send("#{authentication_attribute}=", new_value)
+      else
+        say("Unknown credential field #{k}, ignoring")
+      end
+    end
+    auth.save!
+  end
+
+  def decrypted_awx_value(field, value, pk)
+    require 'awesome_spawn'
+
+    return value unless value.include?("$encrypted$")
+
+    env = {
+      "FIELD_NAME"  => field,
+      "VALUE"       => value,
+      "SECRET_KEY"  => secret_key,
+      "PRIMARY_KEY" => pk
+    }
+    result = AwesomeSpawn.run("python3", :params => [script_path], :env => env)
```

Comment:
First off, I expect this comment be rejected with a "too much work response" but...

Curious if is possible to write the script in a way where we aren't doing a process spawn for every record individual record in the awx auth table.  Might make it more insecure, so not really sure if that is worth it, just curious if that might things a bit more efficient.

Do we want to possibly check with some folks over on the tower team to see if they have a reference database/value available of the max number of auth records they have seen.  I don't expect we would even have someone even close to matching that value using `EmbeddedAnsible` with MIQ, but it would give us an upper bound to consider for this migration.

Reply:
Also, is there any `AwesomeSpawn` logging that we need to be disabling or filtering here?

Reply:
> just curious if that might things a bit more efficient.

I'm sure it would, but like I said with the find in batches comment, I'm not expecting a ton of these.

> Do we want to possibly check with some folks over on the tower team to see if they have a reference database/value available of the max number of auth records they have seen.

Asked

> is there any AwesomeSpawn logging that we need to be disabling or filtering here?

Maybe? It looks like it will log the stdout if the stderr is empty and the script fails, so I guess that could have something sensitive in it.

Reply:
They're saying that they see "from dozens to hundreds right now". So I'm not sure it's worth it given that in a very non-scientific test I'm seeing about 10 decrypts per second locally using this script.